File: unity-package/Editor/Tileable/SeamAnalysis.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace UnityAiAssets.Editor.Tileable
{
    public sealed class SeamAnalysisResult
    {
        public float HorizontalMean;
        public float HorizontalMax;
        public float HorizontalPercentile;
        public float HorizontalScore;
        public float VerticalMean;
        public float VerticalMax;
        public float VerticalPercentile;
        public float VerticalScore;
        public float CombinedScore;

        public string QualityLabel
        {
            get
            {
                if (CombinedScore <= SeamThresholds.ExcellentMax) return "excellent";
                if (CombinedScore <= SeamThresholds.AcceptableMax) return "acceptable";
                return "poor";
            }
        }
    }

    /// <summary>
    /// Deterministic left/right and top/bottom edge mismatch diagnostics.
    /// </summary>
    public static class SeamAnalysis
    {
        public static SeamAnalysisResult Analyze(Color32[] pixels, int width, int height)
        {
            if (pixels == null) throw new ArgumentNullException(nameof(pixels));
            if (width < 2 || height < 2) throw new ArgumentException("Texture must be at least 2x2.");
            if (pixels.Length != width * height) throw new ArgumentException("Pixel buffer size mismatch.");

            var horizontal = new List<float>(height);
            for (var y = 0; y < height; y++)
                horizontal.Add(RgbDistance(pixels[y * width], pixels[y * width + (width - 1)]));

            var vertical = new List<float>(width);
            for (var x = 0; x < width; x++)
                vertical.Add(RgbDistance(pixels[x], pixels[(height - 1) * width + x]));

            EdgeStats(horizontal, out var hMean, out var hMax, out var hPct, out var hScore);
            EdgeStats(vertical, out var vMean, out var vMax, out var vPct, out var vScore);

            return new SeamAnalysisResult
            {
                HorizontalMean = hMean,
                HorizontalMax = hMax,
                HorizontalPercentile = hPct,
                HorizontalScore = hScore,
                VerticalMean = vMean,
                VerticalMax = vMax,
                VerticalPercentile = vPct,
                VerticalScore = vScore,
                CombinedScore = (hScore + vScore) * 0.5f
            };
        }

        static float RgbDistance(Color32 a, Color32 b)
        {
            if (a.a < 8 && b.a < 8) return 0f;
            return (Mathf.Abs(a.r - b.r) + Mathf.Abs(a.g - b.g) + Mathf.Abs(a.b - b.b)) / 3f;
        }

        static void EdgeStats(List<float> distances, out float mean, out float max, out float percentile, out float score)
        {
            if (distances.Count == 0)
            {
                mean = max = percentile = score = 0f;
                return;
            }

            var sum = 0f;
            max = 0f;
            for (var i = 0; i < distances.Count; i++)
            {
                sum += distances[i];
                if (distances[i] > max) max = distances[i];
            }

            mean = sum / distances.Count;
            var ordered = new List<float>(distances);
            ordered.Sort();
            percentile = Percentile(ordered, SeamThresholds.EdgePercentile);
            score = Mathf.Min(1f, mean / SeamThresholds.RgbNormalizer);
        }

        static float Percentile(List<float> sorted, float percentile)
        {
            if (sorted.Count == 0) return 0f;
            if (sorted.Count == 1) return sorted[0];
            var rank = (percentile / 100f) * (sorted.Count - 1);
            var low = (int)rank;
            var high = Mathf.Min(low + 1, sorted.Count - 1);
            var frac = rank - low;
            return sorted[low] * (1f - frac) + sorted[high] * frac;
        }
    }
}

File: src/unity_ai_assets/processing/seam_analysis.py
"""Deterministic, lightweight seam analysis for tileable textures."""

from __future__ import annotations

from dataclasses import dataclass

from PIL import Image

from unity_ai_assets.processing.seam_thresholds import (
    SEAM_EDGE_PERCENTILE,
    SEAM_RGB_NORMALIZER,
    SEAM_SCORE_ACCEPTABLE_MAX,
    SEAM_SCORE_EXCELLENT_MAX,
)


@dataclass(frozen=True, slots=True)
class SeamAnalysisResult:
    """Objective seam diagnostics (not a perceptual guarantee)."""

    horizontal_mean: float
    horizontal_max: float
    horizontal_percentile: float
    horizontal_score: float
    vertical_mean: float
    vertical_max: float
    vertical_percentile: float
    vertical_score: float
    combined_score: float

    @property
    def quality_label(self) -> str:
        if self.combined_score <= SEAM_SCORE_EXCELLENT_MAX:
            return "excellent"
        if self.combined_score <= SEAM_SCORE_ACCEPTABLE_MAX:
            return "acceptable"
        return "poor"


def _as_rgba_pixels(image: Image.Image) -> list[tuple[int, int, int, int]]:
    rgba = image.convert("RGBA")
    return list(rgba.getdata())


def _rgb_distance(a: tuple[int, int, int, int], b: tuple[int, int, int, int]) -> float:
    # Alpha-aware: when both nearly transparent, treat as matching.
    if a[3] < 8 and b[3] < 8:
        return 0.0
    return (abs(a[0] - b[0]) + abs(a[1] - b[1]) + abs(a[2] - b[2])) / 3.0


def _percentile(sorted_values: list[float], percentile: float) -> float:
    if not sorted_values:
        return 0.0
    if len(sorted_values) == 1:
        return sorted_values[0]
    rank = (percentile / 100.0) * (len(sorted_values) - 1)
    low = int(rank)
    high = min(low + 1, len(sorted_values) - 1)
    frac = rank - low
    return sorted_values[low] * (1.0 - frac) + sorted_values[high] * frac


def _edge_stats(distances: list[float]) -> tuple[float, float, float, float]:
    if not distances:
        return 0.0, 0.0, 0.0, 0.0
    mean = sum(distances) / len(distances)
    maximum = max(distances)
    ordered = sorted(distances)
    percentile = _percentile(ordered, SEAM_EDGE_PERCENTILE)
    score = min(1.0, mean / SEAM_RGB_NORMALIZER)
    return mean, maximum, percentile, score


def analyze_seams(image: Image.Image) -> SeamAnalysisResult:
    """Compare left/right and top/bottom edges; return normalized scores in ``[0, 1]``.

    Lower scores indicate closer edge agreement. Scores are objective diagnostics only.
    """
    width, height = image.size
    if width < 2 or height < 2:
        raise ValueError("image must be at least 2x2 for seam analysis")

    pixels = _as_rgba_pixels(image)

    def at(x: int, y: int) -> tuple[int, int, int, int]:
        return pixels[y * width + x]

    horizontal_distances = [_rgb_distance(at(0, y), at(width - 1, y)) for y in range(height)]
    vertical_distances = [_rgb_distance(at(x, 0), at(x, height - 1)) for x in range(width)]

    h_mean, h_max, h_pct, h_score = _edge_stats(horizontal_distances)
    v_mean, v_max, v_pct, v_score = _edge_stats(vertical_distances)
    combined = (h_score + v_score) / 2.0

    return SeamAnalysisResult(
        horizontal_mean=h_mean,
        horizontal_max=h_max,
        horizontal_percentile=h_pct,
        horizontal_score=h_score,
        vertical_mean=v_mean,
        vertical_max=v_max,
        vertical_percentile=v_pct,
        vertical_score=v_score,
        combined_score=combined,
    )
